src/wallet/rpcwallet.cpp

```diff
@@ -5028,39 +5029,38 @@ UniValue z_sendmany(const UniValue& params, bool fHelp)
     int nMinDepth = parseMinconf(DEFAULT_NOTE_CONFIRMATIONS, params, 2, std::nullopt);
 
     // Fee in Zatoshis, not currency format)
-    CAmount nFee = DEFAULT_FEE;
-    if (params.size() > 3) {
-        if (params[3].get_real() == 0.0) {
-            nFee = 0;
-        } else {
-            nFee = AmountFromValue( params[3] );
-        }
+    std::optional<CAmount> nFee;
+    if (params.size() > 3 && params[3].get_real() != -1.0) {
+        CAmount fixedFee = AmountFromValue( params[3] );
 
         // Check that the user specified fee is not absurd.
         // This allows amount=0 (and all amount < DEFAULT_FEE) transactions to use the default network fee
         // or anything less than DEFAULT_FEE instead of being forced to use a custom fee and leak metadata
         if (nTotalOut < DEFAULT_FEE) {
```

Comment:
I think this no longer makes sense, because a tx that pays `DEFAULT_FEE` will only be mined if it fits within the unpaid action limit.

Reply:
Do you think this whole “absurd” check should go away? It seems quite restrictive, and limits the use of near-dust notes … but only applies when an explicit fee is provided by the user.

Reply:
The "absurd" check shouldn't be restrictive at all; it is supposed to only trigger when fees are *very* large.

Reply:
The absurd check I was referring to was just the conditional here. I’m wondering if I should just do something like use `MINIMUM_FEE` instead of `DEFAULT_FEE`, like
```c++
        // Check that the user specified fee is not absurd.
        // This allows amount <= MINIMUM_FEE transactions to pay up to the minimum
        // conventional fee instead of being forced to use a custom fee and leak metadata
        if (nTotalOut < MINIMUM_FEE) {
            if (fixedFee > MINIMUM_FEE) {
                throw JSONRPCError(
                        RPC_INVALID_PARAMETER,
                        strprintf("Small transaction amount %s has fee %s that is greater than the minimum conventional fee %s", FormatMoney(nTotalOut), FormatMoney(fixedFee), FormatMoney(MINIMUM_FEE)));
            }
        } else {
            // Check that the user specified fee is not absurd.
            if (fixedFee > nTotalOut) {
                throw JSONRPCError(
                        RPC_INVALID_PARAMETER,
                        strprintf("Fee %s is greater than the sum of outputs %s and also greater than the minimum conventional fee", FormatMoney(fixedFee), FormatMoney(nTotalOut)));
            }
        }
```
or if I should remove that whole block of code.

The thing that seems odd about this now is that it only applies to user-provided fees, even if they match the conventional fee. So spending, say, three 9kZAT notes and explicitly paying the conventional fee of 15kZAT, will fail as the total out is 12kZAT, which is more than the minimum fee, but less than the specified fee. But paying less of a fee forces the tx into the unpaid actions bucket. And just not specifying a fee at all means the original tx will succeed with the exact same amounts.

If we wanted to keep this logic, it might make sense to move it into wallet_tx_builder, and _always_ calculate the conventional fee (even when the user has specified a fixed fee) and use the computed fee instead of `DEFAULT_FEE` in that conditional.

I.e., it becomes
```c++
        if (nTotalOut < conventionalFee) {
            if (fixedFee > conventionalFee) {
                throw "can’t pay more than the conventional fee (%s) when the tx amount is less than the conventional fee";
            }
        } else if (fixedFee > nTotalOut) {
                throw "can’t pay a fee that is higher than the tx amount (unless the tx amount is lower than the conventional fee)"
        }
```

It becomes a bit more dynamic in that case – re-sending a tx that failed the first time might select different notes, requiring a different fee.

Reply:
There is never any advantage, with the ZIP 317 algorithm, in paying more than four times the conventional fee (because weight_ratio is capped at 4), so let's make that the limit.

Reply:
Moving this into `wallet_tx_builder`, and capping the fee as @daira suggests, makes total sense to me here.